project/src/com/google/daggerquery/executor/models/Query.java

```diff
@@ -79,20 +86,71 @@ public Query(String typeName, String... parameters) {
+   * The connection between nodes is shown with construction '->'. For example, one of the possible paths
+   * may look like this: "com.google.First -> com.google.Second -> com.google.Third".
+   * </ul>
    */
   public List<String> execute(BindingGraph bindingGraph) {
-    if (name.equals("deps")) {
-      String sourceNode = parameters[0];
-      if (!bindingGraph.getAdjacencyListMap().containsKey(sourceNode)) {
-        throw new IllegalArgumentException("Specified source node doesn't exist.");
-      }
+    String source = parameters[0];
 
-      return bindingGraph.getAdjacencyListMap().get(sourceNode).getDependencyList()
-          .stream().map(DependencyProto.Dependency::getTarget).sorted().collect(toList());
+    if (!bindingGraph.getAdjacencyListMap().containsKey(source)) {
+      throw new IllegalArgumentException("Specified source node doesn't exist.");
+    }
+
+    switch (name) {
+      case DEPS_QUERY_NAME:
+        return bindingGraph.getAdjacencyListMap().get(source).getDependencyList()
+            .stream().map(Dependency::getTarget).sorted().collect(toList());
+
+      case ALLPATHS_QUERY_NAME:
+        String target = parameters[1];
+        Set<String> visitedNodes = new HashSet<>();
+        List<List<String>> result = new ArrayList<>();
+        List<String> path = new ArrayList<>();
+
+        findAllPaths(source, target, path, bindingGraph, visitedNodes, result);
+        return result.stream().map(list -> String.join(" -> ", list)).collect(toList());
     }
 
     throw new NotImplementedException();
   }
+
+  /**
+   * Traverses a {@link BindingGraph} starting from {@code source} node.
+   *
+   * <p>Puts all processed nodes in a {@code visitedNodes} set to avoid loops. Constructs a {@code path} which is
+   * an instance of {@link List<String>}. At each recursive level this variable contains a correct path in a graph from
+   * some root node which was passed in the first call and {@code target} node which is the same for all calls.
+   *
+   * <p>As the result fills provided {@link List<List<String>>} {@code allPaths} with all possible paths between
+   * root and target nodes.
+   */
+  private void findAllPaths(String source, String target, List<String> path, BindingGraph bindingGraph,
+                            Set<String> visitedNodes, List<List<String>> allPaths) {
+    path.add(source);
+
+    // If we've already found a path from `source` to `target`, we can stop and not go deeper.
+    if (source.equals(target)) {
+      allPaths.add(new ArrayList<>(path));
+      path.remove(source);
+      return;
+    }
+
+    visitedNodes.add(source);
+    for (Dependency nextNode: bindingGraph.getAdjacencyListMap().get(source).getDependencyList()) {
+      if (visitedNodes.contains(nextNode.getTarget())) {
+        continue;
+      }
+
+      findAllPaths(nextNode.getTarget(), target, path, bindingGraph, visitedNodes, allPaths);
+    }
+
+    visitedNodes.remove(source);
+    path.remove(source);
```

Comment:
The `.remove(Object)` operation on a `List` costs `O(n)`. If you only operate on the two ends, how about using an alternative data structure like an `ArrayDeque`?

Reply:
Oh, yes, exactly! As it behaves as a stack `ArrayDeque` is pretty good, I've added it.